### ISSUE TITLE
A.18.4 (18/2) Should not require "=" operator be "predefined"

### ISSUE THREAD

**Issue by sttaft** [pat-rogers/Ada-202x-WG9-Informal-Review#144]
A.18.4 (18/2) says:

> The predefined "=" operator for type Cursor returns True if both cursors are No_Element, or designate the same element in the same container.

In fact, it is fine if there is a user-defined "=" operator for type Cursor, so long as it composes properly.  Hence, I would suggest we drop the "predefined" and either add an AARM note or an Implementation Requirement that this "=" operator composes properly.

Same problem occurs in A.18.2 (87/2), A.18.3 (59/2), A.18.7 (17/2), and A.18.10 (76/3)

All five of these can be found by searching the RM for the exact phrase "if both cursors"

**Comment by ARG-Editor** [comment]
4.5.2(32.1/1) clearly applies to type Cursor (it is a nonlimited type declared in a language-defined package), so I don't think anything normative about composition is required (it is required of equality of all language-defined types). I won't even bother with an AARM note, since we discuss the fact that language blanket rules apply to containers in the introduction (AARM A.18(5.q/2) specifically calls out composition).

I personally think this request is overly pedantic; from the perspective of a user of a package, whether an operator is predefined or user-defined is irrelevant - whether something is privately overridden is neither visible nor matters. I'd guess that there are quite a few similar wordings throughout the RM. Moreover, this wording has been present since Ada 2005, and there's no evidence that anyone was confused.

Anyway, I'm rather close to clicking on the deferred label, but I'll leave this unmarked for now, if there is time before next Friday, I'll deal with it.

**Comment by sttaft** [comment]
This is not actually just "pedantic" since attempts to ensure that the "predefined" equality had the right properties significantly complicated the implementation, and led to some bugs.  Once we concluded that all that mattered was that the equality operator composed, we were able to simplify things and avoid various bugs.  So I believe we should at a minimum remove "predefined" from the normative wording.  Whether we should also add an AARM note somewhere is a separate issue, and I leave that in the Editor's hands...

**Comment by ARG-Editor** [comment]
I disagree - "as if" changes to language-defined packages are always allowed, and certainly redefining an operator falls into that category. If that wasn't true, pretty much the entire set of Ada predefined units would require heroic efforts to implement. Your always allowed to add things (declarations, context clauses, etc) that don't change the user semantics of a unit. Nothing given in English text changes that fact, so such a concern is pedantic at best. I think there are dozens of similar cases in the RM, and I don't see any reason to prioritize this one over any others (other than that someone is complaining). Still, the absolute lowest priority to fix for me.

**Comment by ARG-Editor** [comment]
I've made a mostly empty AI (AI12-0434-1) for this topic. I don't believe that just dropping "predefined" works, as there is no equality visible in the specification and it is unclear what is being talked about in this case (there could be other equalities that are user-defined or hidden from users). I suppose we could say "primitive" equality here, similar to the rules in 4.5.2, which refers to the equality that is used for composition.
